qip-0009.md

```diff
@@ -0,0 +1,529 @@
+  License: BSD-2-Clause
+ ```
+
+## Abstract
+
+This QIP defines a block interlink structure to be used in Quai as described in
+[1] with few modifications.
+
+## Motivation
+
+To Prove/Verify the Proof Of Work done on Quai since the genesis using a
+non-interactive proof of proof of work (NiPoPoW). Also verify a transaction inclusion in a Block at a depth for which the light node(SPV client) doesn't have the state to verify. This requires using only
+O(log(N)) headers instead of the whole chain history. Specifically in Quai, the plan is to use interlinks for work based
+snap sync and to use nodes as trustless RPC providers.
+
+## Specification
+
+### Definitions
+1. $Order(O)$ - Quai blockchain is merged mined hierarchy of blockchains organized into different levels, Prime, Region and Zone. Prime has a order of 0, Region has order of 1, Zone has order 2. Order of a Block(B) is defined as $O(B)$. $O(B_p)=0$, $O(B_r)=1$, and $O(B_z)=2$. The organization of the levels and merged mining is explained in the [qip0005](qip-0005.md) spec.
+
+### Overview
+
+Interlink data structure is a skip-list that helps a verifier to process a
+sparse subset of the blockchain.
+
+Proof of Work function of a Valid Block in the blockchain satisfies the Mining Target(T) i.e $id \leq T$ . 
+
+Traditional Interlinks are organized into different levels. If $id \leq
+\frac{T}{2^{k \times \mu}}$ we say the block is of rank $\mu$. k is a compression
+parameter and determines the range between interlinks. This is a small
+deviation from the equation in [1] $id \leq \frac{T}{2^{\mu}}$. Having a bigger
+range between interlink levels helps with proof compression while preserving
+the security guarantees.Rank in Quai is analogous to Level in paper[1]. Since Quai has three layers namely Prime, Region, Zone. Prime chain acts as the base chain and has rank of 0. Region chain has a rank of -1 and Zone chain has a rank of -2 by definition. [1] defines $\mu$ for a block $B$ as $$\mu=\log(T)-\log(id(B))$$
+This changes in Quai to $$\mu=\frac{\log(T)-\log(id(B)}{k} $$ For blocks with rank higher than Prime ($\mu=0$).
+
+Genesis Block is set to have $id=0$ and $\mu=\inf$.
+
+Now for the implemention of interlinks in Quai it's established that,
+* Frequency of interlink computation will be on every Prime Block
+* The Target used to calculate the Interlinks will be the  $T_p$. ($T_p$ is the Target used to determine if the block is a prime block).  All interlink hash calculations will be done on this Target.
+
+## Implementation
+
+Given the understanding of this theory, for a given k, and num of levels(L) the interlink data structure would look like this.
+
+```{go}
+type InterLinks []Hash
+```
+
+This QIP proposes to add new field to the header:
+* Merkle root of the interlink. This will be used to store/reference the interlink data structure. This will be stored in the body of the block.
+
+```{go}
+type Header struct {
+  ...
+  ...
+  interlinkRootHash Hash
+  ...
+  ...
+}
+```
```

Comment:
Are we only leaving a hash of the interlink in the header? I thought we decided to put the full structure in the header (saving header bytes is the entire reason why you're using compression factor 'k' instead of building a full binary interlink)

If we're just storing it in the block, then we should not be compressing the structure (get rid of 'k'). **Also, need to add the full structure to the block body, not just the hash in the header.**

Reply:
Why do we need full structure in the block body, there is no requirement for it as long as the hash is verified in each header